lib/internal/test_runner/runner.js

```diff
@@ -446,21 +458,42 @@ function run(options) {
   }
 
   const root = createTestTree({ concurrency, timeout, signal });
-  const testFiles = files ?? createTestFileList();
+  let testFiles = files ?? createTestFileList();
 
   let postRun = () => root.postRun();
   let filesWatcher;
   if (watch) {
+    // Overriding testFiles so we won't try to consume again the stream
+    testFiles = ArrayIsArray(testFiles) ? testFiles : testFiles.toArray();
+
+    // TODO - should not pass a promise
     filesWatcher = watchFiles(testFiles, root, inspectPort, signal, testNamePatterns);
     postRun = undefined;
   }
+
+  const runSingleFile = (path) => {
+    const subtest = runTestFile(path, root, inspectPort, filesWatcher, testNamePatterns);
+    filesWatcher?.runningSubtests.set(path, subtest);
+    return subtest;
+  };
+
   const runFiles = () => {
     root.harness.bootstrapComplete = true;
-    return SafePromiseAllSettledReturnVoid(testFiles, (path) => {
-      const subtest = runTestFile(path, root, inspectPort, filesWatcher, testNamePatterns);
-      filesWatcher?.runningSubtests.set(path, subtest);
-      return subtest;
-    });
+
+    if (isNodeStream(testFiles)) {
+      return testFiles
+        .map(runSingleFile, { __proto__: null, signal, concurrency: NumberMAX_SAFE_INTEGER })
```

Comment:
used this concurrency to not wait for tests otherwise tests that not finish not allow us to run other tests, and this is also to keep the behavior of AllSettled...

will be changed to `Infinity` once the below PR is merged:
- #48588